aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -371,6 +371,144 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+  // of the solution.
+  if (self_batch_stride >= max_stride) {  // This is Case 1
+    HANDLE_EQUATION_CASE(self_batch_, max_, min_)
+  } else if (self_batch_stride > min_stride) {  // This is Case 2
+    HANDLE_EQUATION_CASE(max_, self_batch_, min_)
+  } else {  // This is Case 3
+    HANDLE_EQUATION_CASE(max_, min_, self_batch_)
+  }
+
+  int64_t row, col;
+  if (row_is_max_stride) {
+    row = max_idx;
+    col = min_idx;
+  } else {
+    row = min_idx;
+    col = max_idx;
+  }
+
+  bool mask = upper ? (col - row >= k) : (col - row <= k);
+
+  // Now compute the offset for the result tensor
+  int64_t res_offset = self_batch_idx * res_batch_stride + row * res_row_stride + col * res_col_stride;
+  result[res_offset] = mask ? self[linear_idx] : scalar_t(0);
+}
+
+#define DECLARE_TRIL_TRIU_VARIABLES(SELF, RESULT) \
+  int64_t N = SELF.numel(), \
+          self_batch_stride = SELF.dim() > 2 ? SELF.stride(-3) : matrixStride(SELF), \
+          res_batch_stride = RESULT.dim() > 2 ? RESULT.stride(-3) : matrixStride(RESULT), \
+          self_row_stride = SELF.stride(-2), self_col_stride = SELF.stride(-1), \
+          res_row_stride = RESULT.stride(-2), res_col_stride = RESULT.stride(-1); \
+  bool row_is_max_stride = self_row_stride >= self_col_stride; \
+  int64_t max_stride = row_is_max_stride ? self_row_stride : self_col_stride; \
+  int64_t min_stride = row_is_max_stride ? self_col_stride : self_row_stride; \
+  dim3 dim_block = cuda::getApplyBlock(); \
+  dim3 dim_grid; \
+  AT_CHECK(cuda::getApplyGrid(N, dim_grid, SELF.get_device()), "Unable to get dim grid");
+
+Tensor& tril_cuda_(Tensor &self, int64_t k) {
+  if (self.numel() == 0) {
+    return self;
+  }
+  if (checkZeroStride(self)) self = self.contiguous();
+  DECLARE_TRIL_TRIU_VARIABLES(self, self)
+  AT_DISPATCH_ALL_TYPES_AND_HALF(self.type(), "tril", [&]{
+    triu_tril_kernel<scalar_t, false>
+      <<<dim_grid, dim_block, 0, at::cuda::getCurrentCUDAStream()>>>(
+        self.data<scalar_t>(), self.data<scalar_t>(), k, N,
+        res_batch_stride, self_batch_stride, res_row_stride, res_col_stride,
+        max_stride, min_stride, row_is_max_stride);
+  });
+  return self;
+}
+
+Tensor& tril_cuda_out(Tensor &result, const Tensor& self, int64_t k) {
+  result = at::empty_like(self);
+  if (self.numel() == 0) {
+    return result;
+  }
+  Tensor self_c = checkZeroStride(self) ? self.contiguous() : self;
+  DECLARE_TRIL_TRIU_VARIABLES(self_c, result)
+  AT_DISPATCH_ALL_TYPES_AND_HALF(self.type(), "tril", [&]{
+    triu_tril_kernel<scalar_t, false>
+      <<<dim_grid, dim_block, 0, at::cuda::getCurrentCUDAStream()>>>(
+        result.data<scalar_t>(), self_c.data<scalar_t>(), k, N,
+        res_batch_stride, self_batch_stride, res_row_stride, res_col_stride,
+        max_stride, min_stride, row_is_max_stride);
+  });
+  return result;
+}
+
+Tensor& triu_cuda_(Tensor &self, int64_t k) {
```

Comment:
Is it possible to deduplicated `triu_cuda_` and `triu_cuda_out`?
`triu_cuda_` is almost `triu_cuda_out(self, self, k)` (that's what it was written as in the TH code I think)

Reply:
I'll look into this, thank you.